src/query/execute_statement.c

```diff
@@ -13255,12 +13235,69 @@ do_prepare_insert (PARSER_CONTEXT * parser, PT_NODE * statement)
       return error;
     }
 
+  /* clear any previous error indicator because the rest of do_insert is sensitive to er_errid(). */
+  er_clear ();
+
+  /* fetch the class for instance write purpose */
+  if (!locator_fetch_class (class_->info.name.db_object, DB_FETCH_CLREAD_INSTWRITE))
+    {
+      assert (er_errid () != NO_ERROR);
+      return er_errid ();
+    }
+
+  flag = statement->info.insert.spec->info.spec.flag;
+  if (statement->info.insert.do_replace || statement->info.insert.odku_assignments != NULL)
+    {
+      /* Check to see if the class into which we are inserting is part of an inheritance chain. We do not allow these
+       * statements to be executed in these cases as we might have undefined behavior, such as trying to update a
+       * column that belongs to a child for a duplicate key in the parent table that does not have that column. */
+      int allowed = 0;
+      error = is_replace_or_odku_allowed (class_->info.name.db_object, &allowed);
+      if (error != NO_ERROR)
+	{
+	  return error;
+	}
+
+      if (!allowed)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_REPLACE_ODKU_NOT_ALLOWED, 0);
+	  error = er_errid ();
+	  return error;
+	}
+    }
+
   error = is_server_insert_allowed (parser, statement);
   if (error != NO_ERROR || statement->info.insert.server_allowed != SERVER_INSERT_IS_ALLOWED)
     {
       return error;
     }
 
+  if (statement->info.insert.odku_assignments != NULL)
```

Comment:
@lucianlazar2211 

I'd like to ask you considering two issues on the block.
One is to move the logic to `is_server_insert_allowed` and the other is to free `update`.